docs/architecture/adr-020-protobuf-transaction-encoding.md

```diff
@@ -24,76 +25,186 @@ With this in mind, we will tackle the migration path via two main areas, txs and
-itself. Similar to how we described in [ADR 019](./adr-019-protobuf-state-encoding.md),
-the concrete types will be defined at the application level via Protobuf `oneof`.
+Since interface values are encoded with `google.protobuf.Any` in state (see [ADR 019](adr-019-protobuf-state-encoding.md)),
+ `sdk.Msg`s are encoding with `Any` in transactions.
+ 
+ One of the main goals of using `Any` to encode interface values is to have a
+ core set of types which is reused by apps so that
+ clients can safely be compatible with as many chains as possible.
+ 
+ It is one of the goals of this specification to provide a flexible cross-chain transaction
+ format that can serve a wide variety of use cases without breaking client 
+ compatibility.
+
+In order to facilitate signing, transactions are separated into `TxBody`,
+which will be re-used by `SignDoc` below, and `signatures`:
+
+```proto
+// types/types.proto
+package cosmos_sdk.v1;
+
+message Tx {
+    TxBody body = 1;
+    repeated bytes signatures = 2;
+}
+
+message TxBody {
+    repeated google.protobuf.Any messages = 1;
+    repeated SignerInfo signer_info = 2;
+    Fee fee = 3;
+    string memo = 4;
+    int64 timeout_height = 5;
+    repeated google.protobuf.Any extension_options = 6;
+}
+
+message SignerInfo {
+    PublicKey pub_key = 1;
+    SignMode mode = 2;
+}
+
+enum SignMode {
+    SIGN_MODE_DEFAULT = 0;
+    SIGN_MODE_LEGACY_AMINO = -1;
+    SIGN_MODE_EXTENDED = 1;
+}
+```
+
+As will be discussed below, in order to include as much of the `Tx` as possible
+in the `SignDoc`, `SignerInfo` is separated from signatures so that only the
+raw signatures themselves live outside of `TxBody`.
 
-The application will define a single canonical `Message` Protobuf message
-with a single `oneof` that implements the SDK's `Msg` interface.
+Because we are aiming for be a flexible, extensible cross-chain transaction
+format, new transaction processing options should be added to `TxBody` as soon
+those use cases are discovered, even if they can't be implemented yet.
 
-Example:
+Because there is coordination overhead in this, `TxBody` includes an
+`extension_options` field which can be used for any transaction processing
+options that are not already covered. App developers should, nevertheless,
+attempt to upstream important improvements to `Tx`.
 
-```protobuf
-// app/codec/codec.proto
+### Signing
 
-message Message {
-  option (cosmos_proto.interface_type) = "github.com/cosmos/cosmos-sdk/types.Msg";
+Signatures are structured using the `SignDoc` below which reuses `TxBody` and only
+adds the fields which are needed for signatures but not present on `TxBody`:
 
-  oneof sum {
-    cosmos_sdk.x.bank.v1.MsgSend              msg_send             = 1;
-    cosmos_sdk.x.bank.v1.MsgMultiSend         msg_multi_send       = 2;
-    cosmos_sdk.x.crisis.v1.MsgVerifyInvariant msg_verify_invariant = 3;
-    // ...
-  }
+```proto
+// types/types.proto
+message SignDoc {
+    TxBody body = 1;
+    string chain_id = 2;
```

Comment:
`chain_id` shouldn't get encoded in the transaction (because it's just redundant) but is needed for signing. So it should be in `SignDoc` but not `Tx`

Reply:
Good point, thanks. That's the same that is done in Weave by the way (http://docs.iov.one/docs/iov-name-service/clients/tx-sign-spec).